### ISSUE TITLE
default export handling in CJS and IIFE bundles

### ISSUE THREAD

**Issue by bfred-it** [rollup/rollup#587]
I'm sure this has been talked about before but I can't find it. :-/

The way `default` exports behave when there are additional exports is counterintuitive.

```js
export default function () {}
```

Becomes in IIFE and CJS respectively:

```js
var doStuff = (function () {
	'use strict';
	function index () {}
	return index;
}());
```
```js
'use strict';
function index () {}
module.exports = index;
```

So far so good, even though I wish the code was shorter.

But then when adding a named export:

```js
export default function () {}
export function other() {}
```

Becomes:

```js
'use strict';

function index () {}
function other() {}

exports['default'] = index; // Boom, broken
exports.other = other;
```
```js
(function (exports) {
	'use strict';

	function index () {}
	function other() {}

	exports['default'] = index; // Boom, broken
	exports.other = other;

}((this.doStuff = this.doStuff || {}))); // Also, why this check only here? In some cases the behavior changes completely.
```

---

The core problem is that the if with a single export I can just use
```js
window.doStuff()
```

adding another one it becomes an unintuitive and unideal
```js
window.doStuff['default']()
```

Same goes for the CJS export, which without [this issue](https://github.com/rollup/rollup/issues/496) sorted out, a browserify user would have to use 

```js
var doStuff = require('do-stuff')['default'];
```

Or even

```js
import doStuff from 'do-stuff';
doStuff['default']();
```

---

rollup is an amazing tool to [bundle libraries](https://github.com/frostney/rollup-babel-lib-bundler), but using proper ESM exports with it is not advised and defeats the purpose of the ES6-style rollup output, because I have to do this instead:

```js
function doStuff() {}
Object.defineProperty(doStuff, "other", {
	value:  function () {}
});
export default doStuff;
```

I understand the reasoning behind it, because this isn't possible for example:

```js
var value = 5;
Object.defineProperty(value, "other", {
	value:  function () {}
});
export default value;
```

But it still remains a big issue for, what I believe is, the most common use case (i.e. exporting functions and objects as default). For others perhaps rollup could revert to the current behavior and output a warning.

**Comment by Victorystick** [comment]
@bfred-it I get what you're saying, and I feel your pain, but I don't understand if you have any concrete suggestions for improvements. Can you elaborate? Or do you just want to have an open discussion about these issues?

**Comment by bfred-it** [comment]
The second-last code block was my suggestion for default behavior. The second suggestion was to make the output consistent and not change so much (especially for the IIFE version) when there's more than one export. 

If you feel generous, you could make it an option to avoid defining properties as `default` with `--no-default-extension` and that would reset exports to the unsightly `exports['default'] = something`

As outlined, when the user exports a property-less variable (numbers, strings, …) as `default` export, you could either output a warning saying that the output changed to `--no-default-extension` (to give a nice googlable keyword) or that it *needs* that setting to work (I'd prefer the warning so who doesn't want or can't change the settings can still use it)

**Comment by Rich-Harris** [comment]
Opened #631 to address this. Adding named exports as properties of the object that is the default export is a bad, bad idea – apart from the problem you identified (made worse by the fact that you can't even guarantee whether the default export is a primitive value), it doesn't respect ES2015 semantics, and potentially introduces collisions and other bugs. So we won't do that!

A warning makes sense though, so that's what #631 does – if you have mixed named/default exports and you don't specify the `exportMode` option (which can be `default`, `named` or `none`, and is selected by default using do-what-I-mean logic), Rollup will tell you.

Until ES2015 modules are natively supported everywhere, I recommend doing what you're doing – manually adding properties to the default export, rather than using named exports. There's no real benefit to using named exports alongside default ones unless the entire pipeline is ES2015.

**Comment by bfred-it** [comment]
Thanks for responding! 😊

The benefit would be matching existing CJS behavior:

```js
module.exports = () = 1;
module.exports.secondaryTask = () => 2;
```

While keeping ES2015 named exports for tree-shaking:

```js
export default () = 1;
export const secondaryTask = () => 2;
```

As it is, you either get full CJS support (nobody should have libs that suggest using `['default']` to call their main function) ***or*** treeshaking on ES2015 bundles.

The `exportMode` option sounds good as it is, with the warning. Do you think it has room to have my suggested behavior as a mode, since we're leaving it to the user to pick it?

**Comment by Rich-Harris** [comment]
Ah, yes – I suppose if people are consuming your library via `jsnext:main` then there is an advantage.

Here's an approach you could take: `rollup -c --environment BUILD:cjs && rollup -c --environment BUILD:es6`:

```js
// rollup.config.js

var build = process.env.BUILD;

export default {
  entry: build === 'cjs' ? 'src/entry-cjs.js' : 'src/entry-es6.js',
  dest: build === 'cjs' ? 'dist/bundle.cjs.js' : 'dist/bundle.es6.js',
  format: build
};
```

Then, the appropriate entry file could either to the ES2015 thing, or the attaching-to-the-default-exports thing.

I do wonder if it'd make sense to have a `hybrid` option for `exportMode`. But there are definitely some problems with it (live bindings, for example). For now I'd recommend the dual entry file approach.

**Comment by bfred-it** [comment]
Hmmm I suppose that's a good DIY solution. 

By live bindings you mean this?

```js
export let foo = "bar";
export let get = () => foo;

…
import {foo, get} from "module"
foo = 1;
get(); //-> 1
```

I remember seeing that not working with rollup

**Comment by eventualbuddha** [comment]
@bfred-it you can't modify imported bindings in the importing module, but you can modify exported bindings in the exporting module.

**Comment by Rich-Harris** [comment]
The warning is implemented in 0.26.2, so I'll close this
